### ISSUE TITLE
Double label error pattern has inconsistent screen reader support

### ISSUE THREAD

**Issue by joe-watkins** [jquery-validation/jquery-validation#2143]
In my testing of using two labels that point to the same input field, screen readers are all not on the same page in terms of what is announced for that field. While this is totally valid code it creates barriers for people using screen readers. Screen reader users are able to use the virtual cursor to discover errors but there is no guarantee they will find them, as apposed to when focusing upon an input field with an error and having the error announced with the label.

The default error output for jQuery validate is the double label pattern. I used this pen for a default jQuery validate setup: https://codepen.io/joe-watkins/pen/rJKgmG

``` HTML
<label for="name">Name</label>
<input id="name">
<label class="error">This field is required</label>
```

While an author can change this with settings, I'm hoping a conversation could be started around considering either errors nested with the label by default (all screen readers announce this pattern correctly):

``` HTML
<label for="name">Name <span class="error">This field is required</span></label>
<input id="name">
```

or as an alternative, a span that owns a unique ID along with the error message and `aria-describedby` on the input pointing to this span.

``` HTML
<label for="name">Name</label>
<input id="name" aria-describedby="unique-id-here">
<span class="error" id="unique-id-here">This field is required</span>
```
This recommendation is driven from screen reader testing the following url: https://codepen.io/joe-watkins/pen/qPpQVj

**Screen reader results using the double label pattern:**

JAWS / Internet Explorer
- Reads only 2nd <label> “this field is required” when input is focused

NVDA / FireFox
- Reads both <label> “zip code, this field is required” when input is focused

VoiceOver / Safari
- Reads only 1st <label> “zip code” when input is focused

**Screen reader results using aria-describedby:**

`aria-describedby` = all screen readers read both the label and error when input is focused.

### Your environment
* 1.17.0

### Steps to reproduce
- Navigate to this url: https://codepen.io/joe-watkins/pen/qPpQVj
- Test the double label pattern with JAWS + IE, NVDA + FF, and VoiceOver + Safari

### Expected behaviour
Errors are announced by all screen readers when the input field receives focus.

### Actual behaviour
Errors are not announced by all screen readers consistently when the input field receives focus.

I understand this would be a breaking change to many layouts out there, but would be a huge win for accessibility out of the box in the future :) Many engineers out there use the plugin as-is, without advanced customization, so I'm hoping we could make the default behavior more accessible rather than making folks jump through hoops to achieve it. 

What ya think?

**Comment by joe-watkins** [comment]
@RobJohnston Slick! Such a good form pattern there. Thanks for the link! It is awesome how configurable jQuery Validate is :)

I'm crossing fingers for consideration of similar accessible form validation patterns within the core plugin so that when it comes to the default output, the HTML is more accessible/robust. 

Accessibility out-of-the-box vs. overlays or super complex configuration.

A quicker win / hack could be that when you choose an alternative element for error output e.g. `<span>` a unique ID is added to that span and `aria-describedby` is added to the input pointing to this ID -- probably less of a breaking change there.

**Comment by Arkni** [comment]
Hi @joe-watkins 

Thanks a lot for opening this issue. I'm really interested in making this plugin more accessible.

> The default error output for jQuery validate is the double label pattern. I used this pen for a default jQuery validate setup: https://codepen.io/joe-watkins/pen/rJKgmG

We can improve it as you suggested by adding `aria-describedby` attribute.

From the linked codepen, I see that you used `aria-labelledby` attribute instead of `aria-describedby`, is that a typo or that's the correct pattern?

From MDN, [`aria-describedby`][1] provides more information that the user might need while [`aria-labelledby`][2] provides essential information about an object. Based on these 2 links, I think using `aria-labelledby` make more sense in this case. Wdyt?

> While an author can change this with settings, I'm hoping a conversation could be started around considering either errors nested with the label by default (all screen readers announce this pattern correctly):

That would be a BC break. We can consider doing or adding support for that in the next major version.

> or as an alternative, a span that owns a unique ID along with the error message and aria-describedby on the input pointing to this span.

This pattern is already possible by setting the `errorElement` option to `span`.

Overall, I think for now we should add `aria-describedby`/`aria-labelledby` attributes to the input and point them to  the error label. Does that works for you?

@RobJohnston thanks for the links. Those demos will be a big inspiration for us.

[1]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute
[2]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute

**Comment by RobJohnston** [comment]
@wet-boew/access-working-group

**Comment by joe-watkins** [comment]
Hi @Arkni Thanks so much for the interest and getting back to me. I'd be happy to help by reviewing any commits around this :)

> From the linked codepen, I see that you used aria-labelledby attribute instead of aria-describedby, is that a typo or that's the correct pattern?

In the demo the last example uses `aria-describedby` I had `aria-labelledby` in there for testing purposes. Sorry for the confusion :)

> From MDN, aria-describedby provides more information that the user might need while aria-labelledby provides essential information about an object. Based on these 2 links, I think using aria-labelledby make more sense in this case. Wdyt?

`aria-describedby` is better as it builds upon the `<label>` while `aria-labelledby` could actually overwrite the value of the label. Both approaches could work but I've seen `aria-describedby` used more in validation. Cramming it all into the label is a more robust method to get AT to recognize/announce the error in older non-aria supporting AT but these days, ARIA should do the trick.

> Overall, I think for now we should add aria-describedby/aria-labelledby attributes to the input and point them to the error label. Does that works for you?

Awesome! That would be a massive improvement and make the error messaging more accessible. I think this would really only work for when an author chooses a different element other than the default `<label>` element, which is causing the screen reader bugs.

So an ideal fix would be, if an author chooses to use a different element e.g. span it would automatically add a unique ID to that span, and aria-describedby to the input pointing to this element. The addition of `aria-required="true"` to the required fields would be a nice addition as well. e.g.

``` HTML
<label for="name">Name</label>
<input id="name" aria-describedby="unique-id-here" required aria-required="true">
<span class="error" id="unique-id-here">This field is required</span>
```

**Comment by Arkni** [comment]
@joe-watkins 

> aria-describedby is better as it builds upon the <label> while aria-labelledby could actually overwrite the value of the label. Both approaches could work but I've seen aria-describedby used more in validation. Cramming it all into the label is a more robust method to get AT to recognize/announce the error in older non-aria supporting AT but these days, ARIA should do the trick.

Thanks for the clarification :)

> So an ideal fix would be, if an author chooses to use a different element e.g. span it would automatically add a unique ID to that span, and aria-describedby to the input pointing to this element.

That's already built in. See this fiddle https://jsfiddle.net/Arkni/fz7c6o6y/

> The addition of aria-required="true" to the required fields would be a nice addition as well.

We dropped support for it in version `1.17.0`. See issue #1841 and proposal #2012.

**Comment by joe-watkins** [comment]
@Arkni Ha! Awesome! Good work. Really the work is done now folks just need to know about the option.

Really happy to see that functionality built in. I believe the version my client is using may not have this functionality present, so I'll see about getting them upgraded. Sorry if I wasted any of your time :)

Would you be interested in an addition to the `readme` -- an Accessibility section? This could consist of text copy that speaks to using the custom element for greater screen reader support. This feature is buried a bit :)

**Comment by Arkni** [comment]
@joe-watkins 

> Sorry if I wasted any of your time :)

No worries at all :) I'm here to help as much as I can :)

> Would you be interested in an addition to the readme -- an Accessibility section?

A 100%. You're more than welcome to do that. We appreciate any help :)

**Comment by joe-watkins** [comment]
@Arkni Cool beans! I'll fire up a PR for that README addition in the next few days.
